x/ibc/07-tendermint/types/proposal_handle.go

```diff
@@ -0,0 +1,109 @@
+package types
+
+import (
+	"time"
+
+	tmtypes "github.com/tendermint/tendermint/types"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	clienttypes "github.com/cosmos/cosmos-sdk/x/ibc/02-client/types"
+	"github.com/cosmos/cosmos-sdk/x/ibc/exported"
+)
+
+// CheckProposedHeaderAndUpdateState will try to update the client with the new header if and
+// only if the proposal passes and one of the following two conditions is satisfied:
+// 		1) AllowUpdateAfterExpiry=true and Expire(ctx.BlockTime) = true
+// 		2) AllowUpdateAfterMisbehaviour and IsFrozen() = true
+// In case 2) before trying to update the client, the client will be unfrozen by resetting
+// the FrozenHeight to the zero Height. If AllowUpdateAfterMisbehaviour is set to true,
+// expired clients will also be updated even if AllowUpdateAfterExpiry is set to false.
+// Note, that even if the update happens, it may not be successful. The header may fail
+// validation checks and an error will be returned in that case.
+func (cs ClientState) CheckProposedHeaderAndUpdateState(
+	ctx sdk.Context, cdc codec.BinaryMarshaler, clientStore sdk.KVStore,
+	header exported.Header,
+) (exported.ClientState, exported.ConsensusState, error) {
+	tmHeader, ok := header.(*Header)
+	if !ok {
+		return nil, nil, sdkerrors.Wrapf(
+			clienttypes.ErrInvalidHeader, "expected type %T, got %T", &Header{}, header,
+		)
+	}
+
+	// get consensus state corresponding to client state to check if the client is expired
+	consensusState, err := GetConsensusState(clientStore, cdc, cs.GetLatestHeight())
+	if err != nil {
+		return nil, nil, sdkerrors.Wrapf(
+			err, "could not get consensus state from clientstore at height: %d", cs.GetLatestHeight(),
+		)
+	}
+
+	if cs.IsFrozen() {
```

Comment:
nit, can we use switch for all the conditions here and move the contents to private functions? It's kinda hard to follow the logic, especially the last section were we should be checking for expired client

Reply:
can you do a switch on a struct? Otherwise I'm not sure how to do this

Reply:
I tried doing a switch on the client state but got:
`cannot switch on cs (type ClientState) (struct containing []*ics23.ProofSpec cannot be compared)`

Reply:
```go
switch {
case cs.IsFrozen():
// logic
case cs.AllowUpdateAfterExpiry, !cs.IsExpired(consensusState.Timestamp, ctx.BlockTime()):
// logic
case cs.IsExpired(consensusState.Timestamp, ctx.BlockTime()):
// logic
default:
// error invalid case? 
}
```

Reply:
ah I see, this will be much better. I modified the suggestion above a little in the latest commit, let me know if you think it needs any changes